hi-ml-cpath/src/health_cpath/models/encoders.py

```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
+            compute for memory. Note that this is only supported for some encoders.
+        """
+        assert isinstance(feature_extractor_fn, ResNet), "Expected ResNet model for feature_extractor_fn argument."
+        self.feature_extractor_fn = feature_extractor_fn
+        self.checkpoint_segments_size = checkpoint_segments_size
+        self.bn_momentum = bn_momentum
+        if checkpoint_activations:
+            self._set_batch_norm_momentum()
+
+    def _set_batch_norm_momentum(self) -> None:
+        """Set the momentum of batch norm layers in the ResNet model to avoid running statistics from going out of
+        sync due to activations checkpointing. The forward pass is applied twice which results in double updates of
+        these statistics. We can workaround that by using sqrt of default momentum.
+        """
+        if self.bn_momentum is not None:
+            _momentum = self.bn_momentum
+        else:
+            _momentum = math.sqrt(self.feature_extractor_fn.bn1.momentum)
+            self.bn_momentum = _momentum
+
+        # Set momentum for the first batch norm layer
+        self.feature_extractor_fn.bn1.momentum = _momentum
+
+        def _set_bn_momentum(layer_block: nn.Sequential) -> None:
+            for sub_layer in layer_block:
+                for key, layer in sub_layer._modules.items():
+                    if "bn" in key:
+                        assert isinstance(layer, nn.BatchNorm2d), "Expected BatchNorm2d layer."
+                        layer.momentum = _momentum
+
+        # Fetch all nested batch norm layers and set momentum
+        _set_bn_momentum(self.feature_extractor_fn.layer1)
+        _set_bn_momentum(self.feature_extractor_fn.layer2)
+        _set_bn_momentum(self.feature_extractor_fn.layer3)
+        _set_bn_momentum(self.feature_extractor_fn.layer4)
+
+    def _custom_forward(self, images: torch.Tensor) -> torch.Tensor:
+        """Custom forward pass that uses activation checkpointing to save memory."""
+        segments = self.checkpoint_segments_size
+        first_layers = [
+            self.feature_extractor_fn.conv1,
+            self.feature_extractor_fn.bn1,
+            self.feature_extractor_fn.relu,
+            self.feature_extractor_fn.maxpool,
+        ]
+        images = checkpoint_sequential(first_layers, segments, images)
+        images = checkpoint_sequential(self.feature_extractor_fn.layer1, segments, images)
+        images = checkpoint_sequential(self.feature_extractor_fn.layer2, segments, images)
+        images = checkpoint_sequential(self.feature_extractor_fn.layer3, segments, images)
+        images = checkpoint_sequential(self.feature_extractor_fn.layer4, segments, images)
+
+        images = checkpoint(self.feature_extractor_fn.avgpool, images)
+        images = torch.flatten(images, 1)
+        images = checkpoint(self.feature_extractor_fn.fc, images)
+        return images
+
+
+class Resnet18(ResNetCheckpointingMixin, ImageNetEncoder):
```

Comment:
Maybe choose a different class name? `Resnet18withCheckpointing`?

Reply:
Oh I wouldn't change it as it can be used with and without checkpointing, we can enable checkpointning from the container via encoder params checkpoint_activations (I'll rename it as you suggest ;) )